File: src/offer22.Find_Kth_to_Tail/find_kth_to_tail.cpp
/*
struct ListNode {
	int val;
	struct ListNode *next;
	ListNode(int x) :
			val(x), next(NULL) {
	}
};*/
class Solution {
public:
    ListNode* FindKthToTail(ListNode* pListHead, unsigned int k) {
        if (pListHead == nullptr || k == 0) return nullptr;
        ListNode* node1 = pListHead;
        ListNode* node2 = pListHead;
        for (int i = 0; i < k - 1; i++){
            if (node1->next != nullptr){
                node1 = node1->next;
            }
            else
                return nullptr;
        }
        while (node1->next != nullptr){
            node1 = node1->next;
            node2 = node2->next;
        }
        return node2;
    }
};

File: src/offer22.Find_Kth_to_Tail/find_kth_to_tail.java
/*
public class ListNode {
    int val;
    ListNode next = null;

    ListNode(int val) {
        this.val = val;
    }
}*/
public class Solution {
    public ListNode FindKthToTail(ListNode head,int k) {
        if(head == null || k == 0){ // !!!!!!!!!!
            return null;
        }
        
        ListNode fast = head;
        ListNode slow = head;
        while(k > 1){
            if(fast.next == null){
                return null;         //!!!!!!!!!!!!!!!!
            }
            fast = fast.next;
            k--;
        }
        while(fast != null){
            if(fast.next == null){
                break;
            }
            fast = fast.next;
            slow = slow.next;
        }
        return slow;
    }
}

// 上面的版本是牛客的，最后一个while循环的判断可去掉
//下面这个也可以防止k大于链表长度
/*
public class ListNode {
    int val;
    ListNode next = null;

    ListNode(int val) {
        this.val = val;
    }
}*/
public class Solution {
    public ListNode FindKthToTail(ListNode head,int k) {
        if (head == null) {
            return null;
        }
        ListNode slow = head;
        ListNode fast = head;
        while (k > 0) {
            if (fast == null) {
                return null;
            }
            fast = fast.next;
            k--;
        }
        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }
        return slow;
    }
}
// 下面是lc的
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode getKthFromEnd(ListNode head, int k) {
        if (head == null) {
            return null;
        }
        ListNode slow = head;
        ListNode fast = head;
        while (k > 0) {
            fast = fast.next;
            k--;
        }
        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }
        return slow;
    }
}